src/Pool.sol

```diff
@@ -563,110 +585,141 @@ abstract contract Pool {
+    /// @param lastBalance The balance after the last update of the sender.
+    /// If this is the first update of the sender, pass zero.
+    /// @param currReceivers The list of receivers set in the last update of the sender.
     /// @return withdrawableAmt The withdrawable amount
-    function withdrawable(address senderAddr, Receiver[] calldata currReceivers)
-        public
-        view
-        returns (uint128 withdrawableAmt)
-    {
+    function withdrawable(
+        address senderAddr,
+        uint64 lastUpdate,
+        uint128 lastBalance,
+        Receiver[] calldata currReceivers
+    ) public view returns (uint128 withdrawableAmt) {
         SenderId memory senderId = _senderId(senderAddr);
-        return _withdrawable(senderId, currReceivers);
+        return _withdrawable(senderId, lastUpdate, lastBalance, currReceivers);
     }
 
     /// @notice Returns amount of unsent funds available for withdrawal for the sub-sender
     /// @param senderAddr The address of the sender
     /// @param subSenderId The id of the sender's sub-sender
-    /// @param currReceivers The list of the sub-sender's current receivers.
+    /// @param lastUpdate The timestamp of the last update of the sender.
+    /// If this is the first update of the sender, pass zero.
+    /// @param lastBalance The balance after the last update of the sender.
+    /// If this is the first update of the sender, pass zero.
+    /// @param currReceivers The list of receivers set in the last update of the sender.
     /// @return withdrawableAmt The withdrawable amount
     function withdrawableSubSender(
         address senderAddr,
         uint256 subSenderId,
+        uint64 lastUpdate,
+        uint128 lastBalance,
         Receiver[] calldata currReceivers
     ) public view returns (uint128 withdrawableAmt) {
         SenderId memory senderId = _senderId(senderAddr, subSenderId);
-        return _withdrawable(senderId, currReceivers);
+        return _withdrawable(senderId, lastUpdate, lastBalance, currReceivers);
     }
 
     /// @notice Returns amount of unsent funds available for withdrawal for the sender.
     /// @param senderId The id of the sender
-    /// @param currReceivers The list of the sender's current receivers.
+    /// @param lastUpdate The timestamp of the last update of the sender.
+    /// If this is the first update of the sender, pass zero.
+    /// @param lastBalance The balance after the last update of the sender.
+    /// If this is the first update of the sender, pass zero.
+    /// @param currReceivers The list of receivers set in the last update of the sender.
     /// @return withdrawableAmt The withdrawable amount
-    function _withdrawable(SenderId memory senderId, Receiver[] calldata currReceivers)
-        internal
-        view
-        returns (uint128 withdrawableAmt)
-    {
-        Sender memory sender = _loadSender(senderId);
-        _assertCurrReceivers(sender, currReceivers);
+    function _withdrawable(
```

Comment:
I am currently questioning a bit the sense of the `view` methods like `_withdrawable`.   Before the PR the contract stored the state and performed a calculation based on the state.  Which makes sense.

Now, we need to provide the state to perform the calculation.
Then the question is why the method exits at all, I could perform the witthdrawable calcuation off-chain in a libary and maybe only get the hash from the contract to verify.

On the other hand for other applications on-chain it might be useful.  

One idea would be to make the  formulars in pure methods and move it to a library.

Reply:
It's a good observation and I totally agree! You'll like https://github.com/radicle-dev/radicle-streaming/pull/60 :)